presto-accumulo/src/main/java/com/facebook/presto/accumulo/AccumuloClient.java

```diff
@@ -0,0 +1,1177 @@
+     */
+    private Collection<Range> splitByTabletBoundaries(String tableName, Collection<Range> ranges)
+            throws org.apache.accumulo.core.client.TableNotFoundException, AccumuloException, AccumuloSecurityException
+    {
+        ImmutableSet.Builder<Range> rangeBuilder = ImmutableSet.builder();
+        for (Range r : ranges) {
+            // if start and end key are equivalent, no need to split the range
+            if (r.getStartKey() != null && r.getEndKey() != null
+                    && r.getStartKey().equals(r.getEndKey())) {
+                rangeBuilder.add(r);
+            }
+            else {
+                // Call out to Accumulo to split the range on tablets
+                rangeBuilder.addAll(conn.tableOperations().splitRangeByTablets(tableName, r,
+                        Integer.MAX_VALUE));
+            }
+        }
+        return rangeBuilder.build();
+    }
+
+    /**
+     * Gets the TabletServer hostname for where the given key is located in the given table
+     *
+     * @param table Fully-qualified table name
+     * @param key Key to locate
+     * @return The tablet location, or DUMMY_LOCATION if an error occurs
+     */
+    private Optional<String> getTabletLocation(String table, Key key)
+    {
+        try {
+            // Get the Accumulo table ID so we can scan some fun stuff
+            String tableId = conn.tableOperations().tableIdMap().get(table);
+
+            // Create our scanner against the metadata table, fetching 'loc' family
+            Scanner scan = conn.createScanner("accumulo.metadata", auths);
+            scan.fetchColumnFamily(new Text("loc"));
+
+            // Set the scan range to just this table, from the table ID to the default tablet
+            // row, which is the last listed tablet
+            Key defaultTabletRow = new Key(tableId + '<');
+            Key start = new Key(tableId);
+            Key end = defaultTabletRow.followingKey(PartialKey.ROW);
+            scan.setRange(new Range(start, end));
+
+            Optional<String> location = Optional.empty();
+            if (key == null) {
+                // if the key is null, then it is -inf, so get first tablet location
+                for (Entry<Key, Value> kvp : scan) {
+                    location = Optional.of(kvp.getValue().toString());
+                    break;
+                }
+            }
+            else {
+                // Else, we will need to scan through the tablet location data and find the location
+
+                // Create some text objects to do comparison for what we are looking for
+                Text splitCompareKey = new Text();
+                key.getRow(splitCompareKey);
+                Text scannedCompareKey = new Text();
+
+                // Scan the table!
+                for (Entry<Key, Value> kvp : scan) {
+                    // Get the bytes of the key
+                    byte[] keyBytes = kvp.getKey().getRow().copyBytes();
```

Comment:
If the table is empty, then scanner won't return any results and the `foreach` will be skipped entirely.